src/github.com/getlantern/peerscanner/host.go

```diff
@@ -0,0 +1,407 @@
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/getlantern/cloudflare"
+	"github.com/getlantern/enproxy"
+	"github.com/getlantern/tlsdialer"
+	"github.com/getlantern/withtimeout"
+)
+
+var (
+	// Set a short ttl on DNS entries
+	ttl = 30 * time.Second
+
+	// Test with a period of half the ttl
+	testPeriod = ttl / 2
+
+	// If we haven't had a successul test or reset after this amount of time,
+	// pause testing until receipt of next register call.
+	pauseAfter = 10 * time.Minute
+
+	// Limit how long we're willing to wait for status
+	statusTimeout = ttl * 2
+
+	dialTimeout    = 3 * time.Second // how long to wait on connecting to host
+	requestTimeout = 6 * time.Second // how long to wait for test requests to process
+	proxyAttempts  = 1               // how many times to try a test request before considering host down
+
+	// Sites to use for testing connectivity. WARNING - these should only be
+	// sites with consistent fast response times, around the world, otherwise
+	// tests may time out.
+	testSites = []string{"www.google.com", "www.youtube.com", "www.facebook.com"}
+)
+
+type status struct {
+	online            bool
+	connectionRefused bool
+}
+
+// host is an actor that represents a host entry in CloudFlare and is
+// responsible for checking connectivity to the host and updating CloudFlare DNS
+// accordingly. Once a host has been created, it sticks around ad infinitum.
+// If the host hasn't heard from the real-world host in over 10 minutes, it
+// pauses its processing and only resumes once it hears from the client again.
+type host struct {
+	name        string
+	ip          string
+	record      *cloudflare.Record
+	groups      map[string]*group
+	lastSuccess time.Time
+	lastTest    time.Time
+
+	resetCh      chan string
+	unregisterCh chan interface{}
+	statusCh     chan chan *status
+
+	proxiedClient *http.Client
+}
+
+func (h *host) String() string {
+	return fmt.Sprintf("%v (%v)", h.name, h.ip)
+}
+
+/*******************************************************************************
+ * API for interacting with host
+ ******************************************************************************/
+
+// newHost creates a new host for the given name, ip and optional DNS record.
+func newHost(name string, ip string, record *cloudflare.Record) *host {
+	// Cache TLS sessions
+	clientSessionCache := tls.NewLRUClientSessionCache(1000)
+
+	h := &host{
+		name:         name,
+		ip:           ip,
+		record:       record,
+		resetCh:      make(chan string, 1000),
+		unregisterCh: make(chan interface{}, 1),
+		statusCh:     make(chan chan *status, 1000),
+		proxiedClient: &http.Client{
+			Transport: &http.Transport{
+				Dial: func(network, addr string) (net.Conn, error) {
+					return enproxy.Dial(addr, &enproxy.Config{
+						DialProxy: func(addr string) (net.Conn, error) {
+							return tlsdialer.DialWithDialer(&net.Dialer{
+								Timeout: dialTimeout,
+							}, "tcp", ip+":443", true, &tls.Config{
+								InsecureSkipVerify: true,
```

Comment:
We don't care, we're just checking that there's connectivity to the proxy.  Also, as a general note, anyone can run one of these proxies, so there's nothing meaningful for us to verify anyway.